Retry download_file

While testing #130 there one a single run erroring out  on ubuntu with traceback:

```
test/test_install.py:387: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
.tox/py/lib/python3.9/site-packages/datalad_installer.py:1964: in download
    GitHubClient().download_last_successful_artifact(
.tox/py/lib/python3.9/site-packages/datalad_installer.py:2363: in download_last_successful_artifact
    download_zipfile(archive_download_url, target_dir, headers=self.headers)
.tox/py/lib/python3.9/site-packages/datalad_installer.py:2[471](https://github.com/datalad/datalad-installer/actions/runs/3337677319/jobs/5524327583#step:5:472): in download_zipfile
    zippath = download_to_tempfile(zip_url, suffix=".zip", headers=headers)
.tox/py/lib/python3.9/site-packages/datalad_installer.py:2461: in download_to_tempfile
    download_file(url, p, headers)
.tox/py/lib/python3.9/site-packages/datalad_installer.py:2449: in download_file
    with urlopen(req) as r:
/opt/hostedtoolcache/Python/3.9.15/x64/lib/python3.9/urllib/request.py:214: in urlopen
    return opener.open(url, data, timeout)
/opt/hostedtoolcache/Python/3.9.15/x64/lib/python3.9/urllib/request.py:523: in open
    response = meth(req, response)
/opt/hostedtoolcache/Python/3.9.15/x64/lib/python3.9/urllib/request.py:632: in http_response
    response = self.parent.error(
/opt/hostedtoolcache/Python/3.9.15/x64/lib/python3.9/urllib/request.py:555: in error
    result = self._call_chain(*args)
/opt/hostedtoolcache/Python/3.9.15/x64/lib/python3.9/urllib/request.py:494: in _call_chain
    result = func(*args)
/opt/hostedtoolcache/Python/3.9.15/x64/lib/python3.9/urllib/request.py:747: in http_error_302
    return self.parent.open(new, timeout=req.timeout)
/opt/hostedtoolcache/Python/3.9.15/x64/lib/python3.9/urllib/request.py:517: in open
    response = self._open(req, data)
/opt/hostedtoolcache/Python/3.9.15/x64/lib/python3.9/urllib/request.py:534: in _open
    result = self._call_chain(self.handle_open, protocol, protocol +
/opt/hostedtoolcache/Python/3.9.15/x64/lib/python3.9/urllib/request.py:[494](https://github.com/datalad/datalad-installer/actions/runs/3337677319/jobs/5524327583#step:5:495): in _call_chain
    result = func(*args)
/opt/hostedtoolcache/Python/3.9.15/x64/lib/python3.9/urllib/request.py:1389: in https_open
    return self.do_open(http.client.HTTPSConnection, req,
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <urllib.request.HTTPSHandler object at 0x7ff7d74e7460>
http_class = <class 'http.client.HTTPSConnection'>
req = <urllib.request.Request object at 0x7ff7d74fe4f0>
http_conn_args = {'check_hostname': None, 'context': None}
host = 'pipelines.actions.githubusercontent.com'
h = <http.client.HTTPSConnection object at 0x7ff7d74fe3a0>

    def do_open(self, http_class, req, **http_conn_args):
        """Return an HTTPResponse object for the request, using http_class.
    
        http_class must implement the HTTPConnection API from http.client.
        """
...    
        try:
            try:
                h.request(req.get_method(), req.selector, req.data, headers,
                          encode_chunked=req.has_header('Transfer-encoding'))
            except OSError as err: # timeout error
>               raise URLError(err)
E               urllib.error.URLError: <urlopen error [Errno 104] Connection reset by peer>

/opt/hostedtoolcache/Python/3.9.15/x64/lib/python3.9/urllib/request.py:1349: URLError
```

I think we should retry some number of times (let's say 5) with some arbitrary growing delays between (so spreads through e.g. a minute) since the underlying issue is just intermittent network/server issue and the tool should be robust to those.

--- a/src/datalad_installer.py
+++ b/src/datalad_installer.py
@@ -53,6 +53,7 @@ from typing import (
     Union,
     cast,
 )
+from urllib.error import HTTPError, URLError
 from urllib.request import Request, urlopen
 from zipfile import ZipFile
 
@@ -2445,10 +2446,24 @@ def download_file(
     log.info("Downloading %s", url)
     if headers is None:
         headers = {}
+    delays = iter([1, 2, 6, 15, 36])
     req = Request(url, headers=headers)
-    with urlopen(req) as r:
-        with open(path, "wb") as fp:
-            shutil.copyfileobj(r, fp)
+    while True:
+        try:
+            with urlopen(req) as r:
+                with open(path, "wb") as fp:
+                    shutil.copyfileobj(r, fp)
+            return
+        except URLError as e:
+            if isinstance(e, HTTPError) and e.code not in (500, 502, 503, 504):
+                raise
+            try:
+                delay = next(delays)
+            except StopIteration:
+                raise
+            else:
+                log.warning("Request to %s failed: %s", url, e)
+                log.info("Retrying in %d seconds", delay)
 
 
 def download_to_tempfile(

